From a7fcd4ef6d00479e2631fea09c654557c0d6979d Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Sun, 12 Sep 2021 12:29:58 +0300
Subject: [PATCH] [SW-55224] gaudi2: add get_img_addr for arc

Currently, ARC is using lkd_dev_spec_get_img_addr from the weak
function, which is actually not a correct implementation for ARC.
This patch fix it by adding lkd_dev_spec_get_img_addr specific
to ARC. This function is converting the base address
from ARC view address to AXI (absolute) address. Base address
can reside in spsram or in sram (default option).

Change-Id: I73ecf31da0f480c015daa1eafda3fdc94def1507
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 42d4f73..ae5f24e 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -43,6 +43,20 @@
 #define mmPCIE_WRAP_VUART_TX_DATA       0x4C15F60
 #define mmPCIE_WRAP_VUART_TX_STATUS     0x4C15F64
 
+#define ARC_REL_ADDR_MASK	0x000000000FFFFFFFULL
+#define HL_AXI_GENERAL_BASE	0x1000007FF0000000ULL
+
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	uintptr_t base_offset = base & ARC_REL_ADDR_MASK; /* offset in general region */
+
+	/* SPSRAM fit can only be in general region */
+	if ((base >= SP_SRAM_BASE) && (base < SP_SRAM_BASE + HL_SP_SRAM_SIZE))
+		return base_offset | HL_AXI_GENERAL_BASE;
+
+	return base_offset | SDRAM_BASE_FULL;
+}
+
 void preboot_putc_virt_uart(const char c)
 {
 	/*
